src/Lib/Managers/UndoManager.vala

```diff
@@ -0,0 +1,107 @@
+
+
+
+
+
+
+
+public class Akira.Lib.Managers.UndoManager : Object {
+
+    private GLib.Queue<string> undos;
+    private GLib.Queue<string> redos;
+
+    public UndoManager () {
+        undos = new GLib.Queue<string>();
+        redos = new GLib.Queue<string>();
+    }
+
+
+    public void add_undo(Akira.Lib.Canvas canvas) {
+        redos.clear ();
+        inner_add_undo(Akira.FileFormat.JsonContent.serialize_canvas(canvas));
+        debug("%d %d", (int)undos.get_length(), (int)redos.get_length());
+    }
+
+    public void apply_undo(Akira.Lib.Canvas canvas) {
+        if (undos.get_length () == 0)  {
+            debug("undo empty");
+            return;
+        }
+
+        var old_undo = undos.pop_head();
+        inner_add_redo(old_undo);
+
+        try {
+            var parser = new Json.Parser ();
+            parser.load_from_data (old_undo);
+            var obj = parser.get_root ().get_object ();
+
+            clear_canvas(canvas);
+            Akira.FileFormat.JsonLoader.inner_load_content(canvas, obj);
```

Comment:
I'm not sure about this, it seems a bit extreme.
With this if we move an item and we want to revert to its previous position, we're clearing the canvas and reloading everything. That seems expensive.

Reply:
If clearing and reloading is expensive, we have bigger issues than undo/redo. There is no reason the view should be expensive to regenerate.

I'm open to new ideas, but am cautious about an alternative that doesn't require cleaning the slate. They tend to be rather brittle and crash or artifact prone approaches.

Reply:
To be clear, the current implementation of clearing and reloading is not efficient, but that's a different story.